src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
 
-            // try fold two args without flattening.
-            var folded = TryFoldTwoConcatOperands(loweredLeft, loweredRight);
-            if (folded != null)
+            if (_inExpressionLambda)
             {
-                return folded;
+                // Case 1, expression trees
+                return RewriteStringConcatInExpressionLambda(
+                    originalOperator,
+                    compoundAssignmentLeft: null,
+                    compoundAssignmentOperatorKind: null,
+                    compoundAssignmentSyntax: null,
+                    compoundAssignmentType: null);
             }
 
-            // flatten and merge -  ( expr1 + "A" ) + ("B" + expr2) ===> (expr1 + "AB" + expr2)
-            ArrayBuilder<BoundExpression> leftFlattened = ArrayBuilder<BoundExpression>.GetInstance();
-            ArrayBuilder<BoundExpression> rightFlattened = ArrayBuilder<BoundExpression>.GetInstance();
+            // We'll walk the children in a depth-first order, pull all the arguments out, and then visit them. We'll fold any constant arguments as
+            // we go, pulling them all into a string literal.
+            CollectAndVisitConcatArguments(originalOperator, visitedCompoundAssignmentLeftRead: null, out var arguments);
 
-            FlattenConcatArg(loweredLeft, leftFlattened);
-            FlattenConcatArg(loweredRight, rightFlattened);
+            return RewriteStringConcat(originalOperator.Syntax, arguments);
+        }
 
-            if (leftFlattened.Any() && rightFlattened.Any())
+        private BoundExpression RewriteStringConcat(SyntaxNode originalSyntax, ArrayBuilder<BoundExpression> visitedArguments)
+        {
+            // The strategy of this rewrite is to pull out all the individual arguments of the concat, visit them as appropriate,
+            // and then construct the final concat all in one go. There are a few different lowering patterns that we take:
+            //
+            // 1. If we're in an expression tree, we have to follow strict binary operator rules. Nothing fancy to do, just do a recursive
+            //    rewrite into nested `string.Concat` calls.
+            // 2. If all the added expressions are constants, we can fold them all together and return a single constant.
+            // 3. If all the added expressions are strings, then we want to use one of the `string.Concat(string)`-based overloads: if 4 or less,
+            //    we'll use one of the hardcoded overloads. Otherwise, we'll use `string.Concat(string[])`.
+            // 4. If all the added expressions are strings or chars, we can use the `string.Concat(ReadOnlySpan<char>)`-based overloads. If there are
+            //    more than 4 arguments, we will instead fall back to `string.Concat(string[])`.
+            // 5. If all the added expressions are strings or objects, we'll use the `string.Concat(string)`-based overloads, and call ToString on the
```

Comment:
>If all the added expressions are strings or objects [](http://example.com/codeflow?start=18&length=51)

Perhaps saying "If there are objects among added expressions" will be more accurate